pulpcore/app/models/task.py

```diff
@@ -248,17 +250,15 @@ def set_completed(self):
 
         This updates the :attr:`finished_at` and sets the :attr:`state` to :attr:`COMPLETED`.
         """
-        # Only set the state to finished if it's not already in a complete state. This is
-        # important for when the task has been canceled, so we don't move the task from canceled
-        # to finished.
-        rows = (
-            Task.objects.filter(pk=self.pk)
-            .exclude(state__in=TASK_FINAL_STATES)
-            .update(state=TASK_STATES.COMPLETED, finished_at=timezone.now())
+        # Only set the state to finished if it's running. This is important for when the task has
+        # been canceled, so we don't move the task from canceled to finished.
+        rows = Task.objects.filter(pk=self.pk, state=TASK_STATES.RUNNING).update(
+            state=TASK_STATES.COMPLETED, finished_at=timezone.now()
```

Comment:
What do you think about writing it like this?

```suggestion
        try:
            Task.objects.get(pk=self.pk).update(state=TASK_STATES.COMPLETED, finished_at=timezone.now())
        except DoesNotExist:
            raise RuntimeError(
                _("Task set_completed() occurred but Task {} is not RUNNING.").format(self.pk)
            )
```

Reply:
This is using the `filter().update()` pattern on purpose to avoid race conditions.

Reply:
Maybe to elaborate more, get `get()` and then `update()` will emit a `select` and then an `update` query to the database.
When using `filter().update()` it will emit a single sql statement that selects the items to be updated in the same atomic step. That way we can ensure (on the database layer) that the task can only move from "RUNNING" to "COMPLETED".